### ISSUE TITLE
Missing unique key information after projection

### ISSUE THREAD

**Issue by asenac** [MaterializeInc/materialize#9397]
### What version of Materialize are you using?

From the shell:

```
$ materialized -v

```

Or from the SQL console:

```
materialize=> select mz_version();

```

### How did you install Materialize?

* [ ] Docker image
* [ ] Linux release tarball
* [ ] APT package
* [ ] macOS release tarball
* [ ] Homebrew tap
* [X] Built from source
* [ ] Materialize Cloud

### What was the issue?

```
materialize=> create materialized view v1 as select f1, f2 from t1 group by f1, f2;
CREATE VIEW
materialize=> explain typed select * from v1;
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.v1 (u43)+
 | | types = (integer?, integer?) +
 | | keys = ((#0, #1))            +
 
(1 row)

materialize=> explain typed select f1 from v1 where f1 = f2;
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.v1 (u43)+
 | | types = (integer?, integer?) +
 | | keys = ((#0, #1))            +
 | Filter (#0 = #1)               +
 | | types = (integer, integer)   +
 | | keys = ((#0, #1))            +
 | Project (#0)                   +
 | | types = (integer)            +
 | | keys = ()                    +
 
(1 row)

materialize=>
```

The tuples returned by the last query in the snippet above are guaranteed to be unique, but we fail to report that fact.

Spotted while working on #9354. See https://github.com/MaterializeInc/materialize/pull/9354#issuecomment-984483560

In summary, `Filter` should report `[0]` and `[1]` as its keys, instead of `[0, 1]`.

**Comment by frankmcsherry** [comment]
It looks like this might be an instance of missing functional dependence analysis. If we have a tracking issue for that we could slot this in there, or create one if it is missing. It could be that this is a wrong take, but `f1 = f2` seems like it introduces this, and then we could but do not update the key structure with this functional dependence.

**Comment by asenac** [comment]
The attached PR fixes this issue in a way that is consistent with how the keys information is propagated ATM: as part of the calculation of the type information, which implicitly is doing that FD analysis.
